File: gome-server/src/main/java/com/guardanis/ClientHelper.java
package com.guardanis;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.net.InetAddress;
import java.net.Socket;

import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;

import com.guardanis.SocketManager.ConnectionEvents;
import com.guardanis.commands.Command;
import com.guardanis.gtools.Logger;
import com.guardanis.gtools.gome.Device;

public class ClientHelper extends Thread {
		
	private Socket socket;
	private Device device;
	private ConnectionEvents eventsCallback;
	
    private BufferedReader reader;
	private SocketWriter writer;
	
	private Runnable finishedCallback;
	
	public ClientHelper(Socket socket, final ConnectionEvents eventsCallback){
		this.socket = socket;
		this.eventsCallback = eventsCallback;
		
		Logger.info("Client connected from %1$s, preparing to read.", 
				socket.getInetAddress().getHostAddress());
	}
	
	public ClientHelper setFinishedCallback(Runnable finishedCallback){
		this.finishedCallback = finishedCallback;
		return this;
	}
	
	@Override
	public void run(){
		try{
			reader = new BufferedReader(
					new InputStreamReader(socket.getInputStream()));	
			
			this.device = readDevice();
			
			eventsCallback.onDeviceIdentified(this, device);
			
			writer = new SocketWriter(socket);

			writer.write(InetAddress.getLocalHost()
					.getHostName());
			
			readInputStream();
		}
		catch(Throwable e){ Logger.log(e); }
		
		eventsCallback.onClientDisconnected(this);
		
		if(finishedCallback != null)
			finishedCallback.run();
	}
	
	private Device readDevice() throws Throwable {
		JSONObject deviceData = (JSONObject) new JSONParser()
				.parse(reader.readLine());
		
		return new Device(deviceData);	
	}
	
	private void readInputStream() throws Throwable {		
		String line;		
		
		while ((line = reader.readLine()) != null) 
			process(line);			
	}
	
	private void process(String data) throws Exception {		
//		Logger.info("Data received: " + data);
		
		int delimiter = data.indexOf(":");
		
		Command command = new Command(data.substring(0, delimiter),
				(JSONObject) new JSONParser()
					.parse(data.substring(delimiter + 1)));
		
		device.setLastPing(System.currentTimeMillis());
		
		eventsCallback.onCommandReceived(this, command);
	}
	
	public void write(String data){
		writer.write(data);
	}
	
	public void destroy(){
		try{
			if(reader != null)
				reader.close();
		}
		catch(Exception e){ }

		writer.onDestroy();
		
		try{
			socket.close();
		}
		catch(Exception e){ }
	}
	
	public Device getDevice(){
		return device;
	}

}

File: gome-server/src/main/java/com/guardanis/ClientHelper.kt
package com.guardanis

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.InetAddress
import java.net.Socket

import org.json.simple.JSONObject
import org.json.simple.parser.JSONParser

import com.guardanis.SocketManager.ConnectionEvents
import com.guardanis.commands.Command

class ClientHelper(
        private val socket: Socket,
        private val eventsCallback: ConnectionEvents): Thread() {

    lateinit var device: Device
        private set

    private var reader: BufferedReader? = null
    private var writer: SocketWriter? = null

    private var finishedCallback: Runnable? = null

    init {
        Logger.log("Client connected from ${socket.inetAddress.hostAddress}, preparing to read.")
    }

    fun setFinishedCallback(finishedCallback: Runnable): ClientHelper {
        this.finishedCallback = finishedCallback
        return this
    }

    override fun run() {
        try {
            this.reader = BufferedReader(InputStreamReader(socket.getInputStream()))
            this.device = readDevice()
            this.eventsCallback.onDeviceIdentified(this, device, socket.inetAddress.hostAddress)
            this.writer = SocketWriter(socket)

            writeHostName()

            readInputStream(device)
        }
        catch (e: Throwable) { Logger.log(e) }

        eventsCallback.onClientDisconnected(this)

        finishedCallback?.run()
    }

    @Throws(Throwable::class)
    private fun readDevice(): Device {
        val reader = this.reader ?: throw RuntimeException("Reader cannot be null when requesting device data.")

        val deviceData = JSONParser()
                .parse(reader.readLine()) as? JSONObject ?: throw RuntimeException("Device response was not JSON")

        return Device(deviceData)
    }

    private fun writeHostName() {
        var hostName = "gome-server"

        try {
            hostName = InetAddress.getLocalHost().hostName
        }
        catch (e: Exception) { e.printStackTrace() }

        writer?.write(hostName)
    }

    @Throws(Throwable::class)
    private fun readInputStream(device: Device) {
        val reader = this.reader ?: throw RuntimeException("Reader is null when accessing the input stream!")
        var line: String? = reader.readLine()

        while (line != null) {
            process(device, line)

            line = reader.readLine()
        }
    }

    @Throws(Exception::class)
    private fun process(device: Device, data: String) {
        //		Logger.info("Data received: " + data);

        val delimiter = data.indexOf(":")
        val commandKey = data.substring(0, delimiter)
        val commandData = JSONParser()
                .parse(data.substring(delimiter + 1)) as JSONObject

        val command = Command(commandKey, commandData)

        device.setLastPing(System.currentTimeMillis())

        eventsCallback.onCommandReceived(this, command)
    }

    fun write(data: String) {
        writer?.write(data)
    }

    override fun destroy() {
        reader?.close()
        writer?.onDestroy()

        try {
            socket?.close()
        }
        catch (e: Exception) { }
    }
}
